### PR TITLE
Fix offload extract

### PR DESCRIPTION
**Pull request by dseynhae**
- More comments for maintainability.
- Extract relevant sections up front to speed up the embedded binary search.
- Provide better diagnostics (especially providing info on the result files).
- Fix code to work on Windows (COFF files).

### CODE REVIEW

**Comment by mdtoguchi** [comment]
Looks like there is an issue with the Windows variant of the LIT test.

File: clang/tools/clang-offload-extract/ClangOffloadExtract.cpp [hunk 1]
```diff
@@ -138,105 +137,143 @@ linked fat binary, and store them in separate files.
   }
 
   // We are dealing with an appropriate fat binary;
-  // Locate the section IMAGE_INFO_SECTION_NAME (which contains the
-  // metadata on the embedded binaries)
-  unsigned FileNum = 0;
-
+  // * Create an array all the sections that have
```
Comment (asudarsa) [inline clang/tools/clang-offload-extract/ClangOffloadExtract.cpp:140]:
```suggestion
  // * Create an array of all the sections that have
```

**Review by asudarsa — COMMENTED** [review]
LGTM. Thanks. Please address windows failure.

File: clang/test/Driver/clang-offload-extract.c [hunk 2]
```diff
@@ -1,4 +1,4 @@
-// REQUIRES: x86-registered-target
+// REQUIRES: x86-registered-target
```
Comment (mdtoguchi) [inline clang/test/Driver/clang-offload-extract.c:1]: Looks like you have some trailing whitespace

**Review by mdtoguchi — APPROVED** [review]
OK for driver

**Review by asudarsa — APPROVED** [review]
LGTM. Thanks

**Comment by asudarsa** [comment]
Hi @againull

This change seems good to go. Can you please take a look at this and merge if everything looks ok? The test failure is something we are not able to figure why it's happening. But, it does not seem to be caused by this PR.

Thanks

**Pull request merged**